runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/LockFreeHistogramTest.java

```diff
@@ -0,0 +1,201 @@
+        LockFreeHistogram.Snapshot.create(
+            expectedUnderflow, expectedOverflow, expectedBuckets, bucketType);
+
+    assertThat(snapshot.isPresent(), equalTo(true));
+    assertThat(snapshot.get(), equalTo(expectedSnapshot));
+  }
+
+  @Test
+  public void testUpdate_EmptySnapshot() {
+    HistogramData.BucketType bucketType = HistogramData.LinearBuckets.of(0, 10, 3);
+    LockFreeHistogram histogram =
+        new LockFreeHistogram(KV.of(MetricName.named("name", "namespace"), bucketType));
+    histogram.update(5, 15, 25);
+    Optional<LockFreeHistogram.Snapshot> snapshot_1 = histogram.getSnapshotAndReset();
+
+    assertThat(snapshot_1.isPresent(), equalTo(true));
+
+    Optional<LockFreeHistogram.Snapshot> snapshot_2 = histogram.getSnapshotAndReset();
+    assertThat(snapshot_2.isPresent(), equalTo(false));
+  }
+
+  /** A runnable records 200 values and then calls getSnapshotAndReset. */
+  private static class UpdateHistogramRunnable implements Runnable {
+    private final LockFreeHistogram histogram;
+    private final int val;
+    private final CountDownLatch startSignal;
+    private final CountDownLatch doneSignal;
+    private Optional<LockFreeHistogram.Snapshot> snapshot;
+
+    private static final long valuesRecorded = 200L;
+
+    public UpdateHistogramRunnable(
+        LockFreeHistogram histogram,
+        int val,
+        CountDownLatch startSignal,
+        CountDownLatch doneSignal) {
+      this.histogram = histogram;
+      this.val = val;
+      this.startSignal = startSignal;
+      this.doneSignal = doneSignal;
+      this.snapshot = Optional.empty();
+    }
+
+    @Override
+    public void run() {
+      try {
+        startSignal.await();
+      } catch (InterruptedException e) {
+        return;
+      }
+
+      for (long j = 0; j < valuesRecorded; j++) {
+        histogram.update(val);
+      }
+      snapshot = histogram.getSnapshotAndReset();
+
+      doneSignal.countDown();
+    }
+
+    public long totalCountInSnapshot() {
+      if (snapshot.isPresent()) {
+        return snapshot.get().totalCount();
+      }
+      return 0;
+    }
+
+    public static long numValuesRecorded() {
+      return valuesRecorded;
+    }
+  }
+
+  @Test
+  public void testUpdateAndSnapshots_MultipleThreads() {
+    int numRunnables = 200;
+    CountDownLatch startSignal = new CountDownLatch(1);
+    CountDownLatch doneSignal = new CountDownLatch(numRunnables);
+    ExecutorService executor = Executors.newFixedThreadPool(numRunnables);
+
+    HistogramData.BucketType bucketType = HistogramData.ExponentialBuckets.of(1, 10);
+    LockFreeHistogram histogram =
+        new LockFreeHistogram(KV.of(MetricName.named("name", "namespace"), bucketType));
+
+    UpdateHistogramRunnable[] runnables = new UpdateHistogramRunnable[numRunnables];
+    for (int i = 0; i < numRunnables; i++) {
+      runnables[i] = new UpdateHistogramRunnable(histogram, i, startSignal, doneSignal);
+      executor.execute(runnables[i]);
```

Comment:
you will want to use executor.submit in the loop

or create the array and then use executor.invokeAll() here

call execute 1 at a time will still execute it sequentially on 1 thread since execute blocks